### PR TITLE
wal: fsync directory after wal file rename

### LINKED ISSUE #6368: Possible data loss -- fsync parent directories
I am running a three node etcd cluster. When I insert a new key value pair into the store, I see the following sequence of system calls on the server. 

1 creat("/data/etcd/infra0.etcd/member/wal/0000000000000001-000000000000001b.wal.tmp")
2 append("/data/etcd/infra0.etcd/member/wal/0000000000000001-000000000000001b.wal.tmp")
3 fdatasync("/data/etcd/infra0.etcd/member/wal/0000000000000001-000000000000001b.wal.tmp")
4 rename("/data/etcd/infra0.etcd/member/wal/0000000000000001-000000000000001b.wal.tmp", "/data/etcd/infra0.etcd/member/wal/0000000000000001-000000000000001b.wal")
5 append("/data/etcd/infra0.etcd/member/wal/0000000000000001-000000000000001b.wal")
6 append("/data/etcd/infra0.etcd/member/wal/0000000000000001-000000000000001b.wal")
7 fdatasync("/data/etcd/infra0.etcd/member/wal/0000000000000001-000000000000001b.wal")
8 append("/data/etcd/infra0.etcd/member/wal/0000000000000001-000000000000001b.wal")
9 fdatasync("/data/etcd/infra0.etcd/member/wal/0000000000000001-000000000000001b.wal")
==========Data is durable here -- Ack the user====================

For the 4th operation (rename of wal.tmp to wal) to be persisted to disk, the parent directory has to be fsync'd. If not, a crash just after acknowledging the user can result in a data loss. Specifically, the rename can be reordered on some file systems, thus not issued immediately by the fs. In such a case, on recovery, the server would see the file wal.tmp but not wal. On seeing this, I believe etcd just unlinks the tmp file and therefore can lose the user data. If this happens on two nodes on a three node cluster, then a global data loss is possible.  We have reproduced this particular data loss issue using our testing framework. As a fix, it would be safe to fsync the parent directory on creat or rename of files.

### PR DESCRIPTION
**Pull request by heyitsanthony**
Fixes #6368

### CODE REVIEW

**Comment by xiang90** [comment]
lgtm

File: pkg/fileutil/dir_windows.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+// OpenDir opens a directory in windows with write access for syncing.
+func OpenDir(path string) (*os.File, error) {
+	fd, err := openDir(path)
+	if err != nil {
+		return nil, err
+	}
+	return os.NewFile(uintptr(fd), path), nil
+}
+
+func openDir(path string) (fd syscall.Handle, err error) {
```
Comment (xiang90) [inline pkg/fileutil/dir_windows.go]: this seems to be right. but i am not a window expert...

**Comment by xiang90** [comment]
lgtm

File: pkg/fileutil/dir_windows.go (hunk 1 above)
Comment (heyitsanthony) [inline pkg/fileutil/dir_windows.go]: Yeah, I was testing on a VM to figure this out. The joy of windows.

**Pull request merged**